Added null-pointer guard around setChainParams() calls

bobsummerwill opened the pull request:
And switched asserts to do explicit comparisons against nullptr.

bobsummerwill committed:
Added null-pointer guard around setChainParams() calls, and switched asserts to do explicit comparisons against nullptr.

chriseth commented on libethereum/ChainParams.cpp:
```diff
@@ -116,8 +116,9 @@ ChainParams ChainParams::loadGenesis(string const& _json, h256 const& _stateRoot
-	assert(ret && "Seal engine not found");
-	ret->setChainParams(*this);
+	assert((ret != nullptr) && "Seal engine not found");
+	if (ret != nullptr)
```
The coding style actually recommends `!ret` here.

bobsummerwill replied:
Why?   That is pretty poor advice, IMHO.
And where do you see that recommendation?

chriseth replied:
It is not explicit in the CodingStandards.txt, but it is used like that all over the codebase.
Pro: Much better readable, goes well with type changes
Con: Goes well with type changes

chriseth commented on libethereum/ChainParams.cpp:
```diff
@@ -116,8 +116,9 @@ ChainParams ChainParams::loadGenesis(string const& _json, h256 const& _stateRoot
+		ret->setChainParams(*this);
 	if (sealRLP.empty())
 	{
 		sealFields = ret->sealFields();
```
You use ret again here, so probably better to return early or throw.

chfast commented on libethcore/SealEngine.cpp:
```diff
@@ -51,7 +51,8 @@ void SealEngineFace::verifyTransaction(ImportRequirements::value _ir, Transactio
 	SealEngineFace* ret = create(_params.sealEngineName);
-	assert(!!ret && "Seal engine not found.");
-	ret->setChainParams(_params);
+	assert(ret && "Seal engine not found.");
```
The assert is definitely a bad choice here if `create()` can fail.

chfast commented on libethcore/SealEngine.cpp:
```diff
@@ -51,7 +51,8 @@ void SealEngineFace::verifyTransaction(ImportRequirements::value _ir, Transactio
-	assert(!!ret && "Seal engine not found.");
-	ret->setChainParams(_params);
+	assert(ret && "Seal engine not found.");
+	if (ret)
```
I would prefer using "quick return" pattern as in other file.

chfast commented on libethereum/ChainParams.cpp:
```diff
@@ -117,6 +117,8 @@ SealEngineFace* ChainParams::createSealEngine()
 {
 	SealEngineFace* ret = SealEngineRegistrar::create(sealEngineName);
 	assert(ret && "Seal engine not found");
+	if (!ret)
```
The same issue with assert.

chfast commented on libethereum/ChainParams.cpp:
```diff
@@ -117,6 +117,8 @@ SealEngineFace* ChainParams::createSealEngine()
 	SealEngineFace* ret = SealEngineRegistrar::create(sealEngineName);
 	assert(ret && "Seal engine not found");
+	if (!ret)
+		return NULL;
```
~~NULL~~nullptr.

bobsummerwill commented:
RE:   Quick return.

Original code prior to @chriseth assertion:

```
SealEngineFace* ret = create(_params.sealEngineName);
ret->setChainParams(_params);
return ret;
```

Current code:

```
SealEngineFace* ret = create(_params.sealEngineName);
assert(ret && "Seal engine not found.");
ret->setChainParams(_params);
return ret;
```

My proposed code:

```
SealEngineFace* ret = create(_params.sealEngineName);
assert(ret && "Seal engine not found.");
if (ret)
    ret->setChainParams(_params);
return ret;
```

So what are you suggesting, @chfast?   This?   Is it any better?

```
SealEngineFace* ret = create(_params.sealEngineName);
assert(ret && "Seal engine not found.");
if (!ret)
    return NULL;
ret->setChainParams(_params);
return ret;
```

chfast commented:
In general I propose to remove the asserts. The rest is about coding style only.

bobsummerwill commented:
NOTE - So we had nothing.    Then we added an assert for better error messages in debug builds.   And now I am attempting to NOT crash.      If we really want to follow through here, we should add handling in the sites which call this too.    I've not looked at all those yet.    I'm just trying to improve the code we just modified, so we can get at least these functions in a state which we like.

bobsummerwill commented:
OK - over to you @chriseth :-)

chfast commented:
If asserts are there temporary for diagnostics then add a comment about that and leave them. And add an issue to track the bug later one.

chriseth commented:
So in order to fix this properly, we need to check the call sites.
I would also prefer to add an assert to guard against future sloppy usage.
This is not about finding the cause of a problem but rather about failing fast if the problem occurs or persists.

bobsummerwill commented:
Yes - having the assert AND the runtime check is not a bad thing.

Can we get this committed as-is, please, @chriseth?    It will make things better.

chfast commented:
Just replace `NULL` with `nullptr`.

bobsummerwill commented:
DONE

chfast commented:
So I'm ok with the change. Not sure if you need more confirmations.
